File: src/lib/events.ts
/**
 * Event names emitted by the Rust core, and a typed `listen` helper.
 *
 * `DICTATION_STATE` is the authoritative one the UI renders from; the
 * lifecycle events exist for anything that needs a specific moment.
 */
import { listen, type UnlistenFn } from "@tauri-apps/api/event";
import type {
  DownloadProgress, DictationState } from "./types";

export const EVENTS = {
  dictationState: "dictation:state",
  dictationLevel: "dictation:level",
  dictationStarted: "dictation:started",
  dictationStopped: "dictation:stopped",
  transcriptionStarted: "transcription:started",
  transcriptionComplete: "transcription:complete",
  transcriptionFailed: "transcription:failed",
  processingStarted: "processing:started",
  processingComplete: "processing:complete",
  insertionStarted: "insertion:started",
  insertionComplete: "insertion:complete",
  insertionFailed: "insertion:failed",
  historyChanged: "history:changed",
  settingsChanged: "settings:changed",
  navigate: "navigate",
  modelProgress: "model:progress",
  modelComplete: "model:complete",
  modelFailed: "model:failed",
} as const;

export interface LevelPayload {
  level: number;
}

export interface StartedPayload {
  targetApp: string | null;
}

export interface FailurePayload {
  message: string;
  retryable: boolean;
}

export interface TextPayload {
  text: string;
}

interface EventMap {
  [EVENTS.dictationState]: DictationState;
  [EVENTS.dictationLevel]: LevelPayload;
  [EVENTS.dictationStarted]: StartedPayload;
  [EVENTS.dictationStopped]: null;
  [EVENTS.transcriptionStarted]: null;
  [EVENTS.transcriptionComplete]: TextPayload;
  [EVENTS.transcriptionFailed]: FailurePayload;
  [EVENTS.processingStarted]: null;
  [EVENTS.processingComplete]: TextPayload;
  [EVENTS.insertionStarted]: null;
  [EVENTS.insertionComplete]: null;
  [EVENTS.insertionFailed]: FailurePayload;
  [EVENTS.historyChanged]: null;
  [EVENTS.settingsChanged]: null;
  [EVENTS.navigate]: string;
  [EVENTS.modelProgress]: DownloadProgress;
  /** The model id that finished. */
  [EVENTS.modelComplete]: string;
  [EVENTS.modelFailed]: { modelId: string; message: string };
}

export function on<K extends keyof EventMap>(
  name: K,
  handler: (payload: EventMap[K]) => void,
): Promise<UnlistenFn> {
  return listen<EventMap[K]>(name, (event) => handler(event.payload));
}

File: src-tauri/src/dictation/events.rs
//! The event contract between Rust and the UI.
//!
//! Two layers, on purpose:
//!
//! * `STATE` carries the whole `DictationState` and is what the UI actually
//!   renders from. One authoritative value means the interface cannot show a
//!   contradictory combination of flags.
//! * The lifecycle events mirror the pipeline stages for anything that needs
//!   to react to a specific moment (the HUD's success chime, a future menu-bar
//!   animation) without diffing state.

use serde::Serialize;
use tauri::{AppHandle, Emitter};

use super::machine::DictationState;

pub const STATE: &str = "dictation:state";
pub const LEVEL: &str = "dictation:level";

pub const DICTATION_STARTED: &str = "dictation:started";
pub const DICTATION_STOPPED: &str = "dictation:stopped";

pub const TRANSCRIPTION_STARTED: &str = "transcription:started";
pub const TRANSCRIPTION_COMPLETE: &str = "transcription:complete";
pub const TRANSCRIPTION_FAILED: &str = "transcription:failed";
/// Emitted when a substitute engine served the transcription. A fallback is
/// never silent — see `dictation::fallback`.
pub const TRANSCRIPTION_FELL_BACK: &str = "transcription:fell-back";

pub const PROCESSING_STARTED: &str = "processing:started";
pub const PROCESSING_COMPLETE: &str = "processing:complete";

pub const INSERTION_STARTED: &str = "insertion:started";
pub const INSERTION_COMPLETE: &str = "insertion:complete";
pub const INSERTION_FAILED: &str = "insertion:failed";

/// A new row landed in history; the dashboard and history view refresh.
pub const HISTORY_CHANGED: &str = "history:changed";

/// Settings changed in the backend (e.g. a shortcut re-registration).
pub const SETTINGS_CHANGED: &str = "settings:changed";

#[derive(Clone, Serialize)]
#[serde(rename_all = "camelCase")]
pub struct LevelPayload {
    /// 0.0..=1.0 microphone RMS.
    pub level: f32,
}

#[derive(Clone, Serialize)]
#[serde(rename_all = "camelCase")]
pub struct TextPayload {
    pub text: String,
}

#[derive(Clone, Serialize)]
#[serde(rename_all = "camelCase")]
pub struct FailurePayload {
    pub message: String,
    /// Whether the UI should offer Retry.
    pub retryable: bool,
}

#[derive(Clone, Serialize)]
#[serde(rename_all = "camelCase")]
pub struct StartedPayload {
    /// The app that will receive the text, for the HUD's context line.
    pub target_app: Option<String>,
}

pub fn emit_state(app: &AppHandle, state: &DictationState) {
    if let Err(error) = app.emit(STATE, state) {
        tracing::warn!(?error, "could not emit dictation state");
    }
}

pub fn emit<T: Serialize + Clone>(app: &AppHandle, name: &str, payload: T) {
    if let Err(error) = app.emit(name, payload) {
        tracing::warn!(?error, name, "could not emit event");
    }
}

pub fn emit_bare(app: &AppHandle, name: &str) {
    emit(app, name, ());
}

#[derive(Clone, Serialize)]
#[serde(rename_all = "camelCase")]
pub struct FallbackPayload {
    pub failed_provider: String,
    /// The display name of whatever actually ran, for the HUD.
    pub used_provider: String,
    pub used_model: String,
}
